ui/StatusQ/src/modelentry.cpp

```diff
@@ -0,0 +1,346 @@
+        if(!m_index.isValid())
+        {
+            // Resetting just to cover cases where the rows are removed after the layout change
+            resetItem();
+        }
+        setRow(m_index.row());
+    });
+
+    emit sourceModelChanged();
+}
+
+void ModelEntry::setKey(const QString& key)
+{
+    if(m_key == key) return;
+
+    m_key = key;
+    resetIndex();
+    emit keyChanged();
+}
+
+void ModelEntry::setValue(const QVariant& value)
+{
+    if(m_value == value) return;
+
+    m_value = value;
+    resetIndex();
+    emit valueChanged();
+}
+
+void ModelEntry::setIndex(const QModelIndex& index)
+{
+    if(m_index == index) return;
+
+    m_index = index;
+
+    tryItemResetOrUpdate();
+    setRow(m_index.row());
+}
+
+void ModelEntry::setAvailable(bool available)
+{
+    if(available == m_available) return;
+
+    m_available = available;
+    emit availableChanged();
+}
+
+void ModelEntry::setRoles(const QStringList& roles)
+{
+    if(m_roles.size() == roles.size() && !m_roles.empty() &&
+       std::all_of(roles.begin(), roles.end(), [this](const QString& role) { return m_roles.contains(role); }))
+        return;
+
+    m_roles = roles;
+    emit rolesChanged();
+}
+
+void ModelEntry::setRow(int row)
+{
+    if(m_row == row) return;
+
+    m_row = row;
+    emit rowChanged();
+}
+
+void ModelEntry::setCacheOnRemoval(bool cacheOnRemoval)
+{
+    if(m_cacheOnRemoval == cacheOnRemoval) return;
+
+    resetCachedItem();
+
+    m_cacheOnRemoval = cacheOnRemoval;
+    emit cacheOnRemovalChanged();
+}
+
+void ModelEntry::setItemRemovedFromModel(bool itemRemovedFromModel)
+{
+    if(m_itemRemovedFromModel == itemRemovedFromModel) return;
+
+    m_itemRemovedFromModel = itemRemovedFromModel;
+    emit itemRemovedFromModelChanged();
+}
+
+QModelIndex ModelEntry::findIndexInRange(int start, int end, const QList<int>& roles) const
+{
+    if(!m_sourceModel || m_key.isEmpty()) return {};
+
+    auto keysForRole = m_sourceModel->roleNames().keys(m_key.toUtf8());
+
+    // no matching roles found
+    if(keysForRole.isEmpty() || (!roles.isEmpty() && !roles.contains(keysForRole.first()))) return {};
+
+    for(int i = start; i < end; i++)
+    {
+        auto index = m_sourceModel->index(i, 0);
+        auto data = index.data(keysForRole.first());
+        if(data == m_value) return index;
+    }
+
+    return {};
+}
+
+void ModelEntry::resetIndex()
+{
+    auto index = QModelIndex();
+    if(m_sourceModel) index = findIndexInRange(0, m_sourceModel->rowCount());
+
+    setIndex(index);
+}
+
+void ModelEntry::tryItemResetOrUpdate()
+{
+    if(!m_index.isValid() || !itemHasCorrectRoles())
+    {
+        resetItem();
+        return;
+    }
+
+    updateItem();
+    setAvailable(true);
+}
+
+void ModelEntry::resetItem()
+{
+    // Signal order is important here
+    if(!m_index.isValid())
+    {
+        setAvailable(false);
+    }
+
+    m_item = new QQmlPropertyMap(this);
```

Comment:
as it's `QPointer` offering only automatic clearance, the previous object is not released. Every reset adds new `QQmlPropertyMap` which has lifetime of `this`.